File: python/tests/identity_users_test.py
"""Identity (S-ID) and Users (S-US) acceptance scenarios."""

from __future__ import annotations

from harness.fixture import Fixture
from tests.helpers import Client, client_for, seed_user


def test_s_id_01_missing_x_user_id_is_401(fixture: Fixture) -> None:
    client_for(fixture, None).get("/feed").expect(401).expect_code("identity:missing")


def test_s_id_02_unknown_x_user_id_is_401(fixture: Fixture) -> None:
    client_for(fixture, "nosuchuser").get("/feed").expect(401).expect_code("identity:unknown")


def test_s_us_01_create_user_valid(fixture: Fixture) -> None:
    response = client_for(fixture, None).post(
        "/users", {"handle": "alice", "displayName": "Alice"}
    )
    response.expect(201)
    body = response.json()
    assert body["handle"] == "alice"
    assert body["displayName"] == "Alice"
    assert body["id"]
    assert body["createdAt"]


def test_s_us_02_duplicate_handle_is_409(fixture: Fixture) -> None:
    seed_user(fixture, "bob")
    client_for(fixture, None).post(
        "/users", {"handle": "bob", "displayName": "Other Bob"}
    ).expect(409).expect_code("user:handle:taken")


def test_s_us_03_invalid_handle_is_422(fixture: Fixture) -> None:
    unauth: Client = client_for(fixture, None)
    for handle in ("ab", "UPPER", "has space"):
        unauth.post("/users", {"handle": handle, "displayName": "X"}).expect(
            422
        ).expect_code("user:handle:invalid")


def test_s_us_04_invalid_display_name_is_422(fixture: Fixture) -> None:
    unauth = client_for(fixture, None)
    unauth.post("/users", {"handle": "carol", "displayName": ""}).expect(422).expect_code(
        "user:display_name:invalid"
    )
    unauth.post("/users", {"handle": "carol2", "displayName": "x" * 65}).expect(
        422
    ).expect_code("user:display_name:invalid")


def test_s_us_05_get_existing_user(fixture: Fixture) -> None:
    user = seed_user(fixture, "dave")
    client_for(fixture, user.id).get(f"/users/{user.id}").expect(200)


def test_s_us_06_get_unknown_user_is_404(fixture: Fixture) -> None:
    user = seed_user(fixture, "erin")
    client_for(fixture, user.id).get("/users/nope").expect(404).expect_code("user:not_found")

File: go/tests/identity_users_test.go
package relaytest

import (
	"net/http"
	"testing"
)

// ---- Identity (S-ID) ----

func TestS_ID_01_missing_user_id_is_401(t *testing.T) {
	reset(t)
	newClient(t, "").get("/channels").expect(http.StatusUnauthorized).expectCode("identity:missing")
}

func TestS_ID_02_unknown_user_id_is_401(t *testing.T) {
	reset(t)
	newClient(t, "nobody-0000000").get("/channels").expect(http.StatusUnauthorized).expectCode("identity:unknown")
}

// ---- Users (S-US) ----

func TestS_US_01_create_user_valid(t *testing.T) {
	reset(t)
	resp := newClient(t, "").post("/users", map[string]string{"handle": "ada_lovelace", "displayName": "Ada"})
	resp.expect(http.StatusCreated)
	var u struct {
		ID, Handle, DisplayName, CreatedAt string
	}
	resp.decode(&u)
	if u.Handle != "ada_lovelace" || u.DisplayName != "Ada" {
		t.Fatalf("echo mismatch: %+v", u)
	}
	if u.ID == "" || u.CreatedAt == "" {
		t.Fatalf("id/createdAt missing: %+v", u)
	}
}

func TestS_US_02_duplicate_handle_is_409(t *testing.T) {
	reset(t)
	newClient(t, "").post("/users", map[string]string{"handle": "dup", "displayName": "One"}).expect(http.StatusCreated)
	newClient(t, "").post("/users", map[string]string{"handle": "dup", "displayName": "Two"}).
		expect(http.StatusConflict).expectCode("user:handle:taken")
}

func TestS_US_03_invalid_handle_is_422(t *testing.T) {
	reset(t)
	for _, bad := range []string{"ab", "UPPER", "has space"} {
		newClient(t, "").post("/users", map[string]string{"handle": bad, "displayName": "X"}).
			expect(http.StatusUnprocessableEntity).expectCode("user:handle:invalid")
	}
}

func TestS_US_04_invalid_display_name_is_422(t *testing.T) {
	reset(t)
	long := make([]byte, 65)
	for i := range long {
		long[i] = 'a'
	}
	newClient(t, "").post("/users", map[string]string{"handle": "emptyname", "displayName": ""}).
		expect(http.StatusUnprocessableEntity).expectCode("user:display_name:invalid")
	newClient(t, "").post("/users", map[string]string{"handle": "longname", "displayName": string(long)}).
		expect(http.StatusUnprocessableEntity).expectCode("user:display_name:invalid")
}

func TestS_US_05_get_existing_user(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	newClient(t, ada.ID).get("/users/" + ada.ID).expect(http.StatusOK)
}

func TestS_US_06_get_unknown_user_is_404(t *testing.T) {
	reset(t)
	ada := seedUser(t, "ada")
	newClient(t, ada.ID).get("/users/UNKNOWN0000000").expect(http.StatusNotFound).expectCode("user:not_found")
}
